analyzers/src/SonarAnalyzer.CSharp/Rules/ImplementISerializableCorrectly.cs

```diff
@@ -53,53 +53,43 @@ public class ImplementISerializableCorrectly : SonarDiagnosticAnalyzer
                     implementationErrors.AddRange(CheckGetObjectData(typeDeclarationSyntax, typeSymbol, getObjectData));
-
-                    if (implementationErrors.Count > 0)
+                    if (implementationErrors.Any())
                     {
-                        c.ReportIssue(Diagnostic.Create(Rule,
-                                                                       typeIdentifier.GetLocation(),
-                                                                       additionalLocations: implementationErrors.ToAdditionalLocations(),
-                                                                       properties: implementationErrors.ToProperties()));
+                        c.ReportIssue(Diagnostic.Create(Rule, typeDeclarationSyntax.Identifier.GetLocation(), implementationErrors.ToAdditionalLocations(), implementationErrors.ToProperties()));
                     }
                 },
                 SyntaxKind.ClassDeclaration,
-                SyntaxKindEx.RecordClassDeclaration);
+                SyntaxKindEx.RecordClassDeclaration,
+                SyntaxKindEx.RecordStructDeclaration,
+                SyntaxKind.StructDeclaration);
 
         private static IEnumerable<SecondaryLocation> CheckSerializableAttribute(SyntaxToken typeKeyword, INamedTypeSymbol typeSymbol)
         {
             if (!HasSerializableAttribute(typeSymbol) && !typeSymbol.IsAbstract)
             {
-                yield return new SecondaryLocation(typeKeyword.GetLocation(),
-                                                   $"Add 'System.SerializableAttribute' attribute on '{typeSymbol.Name}' because it implements 'ISerializable'.");
+                yield return new(typeKeyword.GetLocation(), $"Add 'System.SerializableAttribute' attribute on '{typeSymbol.Name}' because it implements 'ISerializable'.");
             }
         }
 
-        // symbol should be checked for null in the caller.
+        // Symbol should be checked for null in the caller.
         private static IEnumerable<TSyntax> DeclarationOrImplementation<TSyntax>(TypeDeclarationSyntax typeDeclaration, IMethodSymbol symbol) =>
-            symbol.PartialImplementationPart != null
-                   && symbol.PartialImplementationPart.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax() is { } partialImplementation
-                   && typeDeclaration.DescendantNodes().Any(x => x.Equals(partialImplementation))
+            symbol.PartialImplementationPart is not null
+            && symbol.PartialImplementationPart.DeclaringSyntaxReferences.First().GetSyntax() is var partialImplementation
+            && typeDeclaration.DescendantNodes().Any(x => x.Equals(partialImplementation))
                 ? new[] { partialImplementation }.Cast<TSyntax>()
-                : symbol.DeclaringSyntaxReferences.Select(r => r.GetSyntax()).Cast<TSyntax>();
+                : symbol.DeclaringSyntaxReferences.Select(x => x.GetSyntax()).Cast<TSyntax>();
```

Comment:
I tried to extract common logic around `PartialImplementationPart` in #5603.

This part cannot use that, because it does something very specific to this rule. This piece of code is not reusable by any other rule and is quite weird (related to not-reporting on specific secondary locations). So I'll leave this one as is.